deployment-server/app/Main.hs

```diff
@@ -0,0 +1,67 @@
+{-# LANGUAGE DeriveAnyClass    #-}
+{-# LANGUAGE DeriveGeneric     #-}
+{-# LANGUAGE NamedFieldPuns    #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards   #-}
+
+module Main where
+
+import           Control.Concurrent       (forkIO)
+import           Control.Concurrent.Chan  (newChan)
+import           Control.Newtype.Generics (Newtype, unpack)
+import           Data.Aeson               (FromJSON, eitherDecodeFileStrict, parseJSON, withObject, (.:))
+import qualified Data.ByteString.Char8    as BS
+import           Data.Map                 as Map
+import           Data.Map                 (Map)
+import           Data.Text                (Text)
+import qualified Data.Text                as Text
+import           Deploy.Server            (app)
+import           Deploy.Worker            (Deployment, SlackChannel, runWorker)
+import           GHC.Generics             (Generic)
+import           Network.Wai.Handler.Warp (run)
+import           Options.Generic          (ParseField, ParseFields, ParseRecord, getRecord)
+import           Servant.GitHub.Webhook   (gitHubKey)
+import           System.IO                (BufferMode (LineBuffering), hSetBuffering, stderr, stdout)
+import           Web.Slack                (mkSlackConfig)
+
+newtype SlackToken = SlackToken Text
+    deriving (Generic, Newtype)
+
+newtype WebhookKey = WebhookKey Text
+    deriving (Generic, Newtype)
+
+data Options = Options
+    { port           :: Int
+    , configDir      :: FilePath
+    , stateFile      :: FilePath
+    , include        :: [String]
+    , keyfile        :: FilePath
+    , slackChannel   :: SlackChannel
+    , deploymentName :: Deployment
+    } deriving (Generic, Show, ParseRecord)
+
+data Secrets = Secrets
+    { githubWebhookKey :: WebhookKey
+    , slackToken       :: SlackToken
+    }
+
+instance FromJSON Secrets where
+    parseJSON =
+        withObject "Secrets" $ \v ->
+            Secrets <$> (WebhookKey <$> v .: "githubWebhookKey") <*> (SlackToken <$> v .: "slackToken")
+
+main :: IO ()
+main = do
+    -- When using systemd, journald does something weird with buffering so lets force it to be line buffered
+    hSetBuffering stdout LineBuffering
+    hSetBuffering stderr LineBuffering
+    Options {..} <- getRecord "Plutus CD Server"
+    eJSON <- eitherDecodeFileStrict keyfile
+    case eJSON of
+        Left err -> putStrLn $ "failed to load keyfile with error " <> err
```

Comment:
As this is running under journald, is it a good idea to exit here with a non-zero code?

Reply:
(Also you'll get bonus points for `hPutStrLn stderr` on the error message.)

Reply:
ok, exit 2